File: Google/059_BasicCalc2.py
# https://leetcode.com/problems/basic-calculator-ii/

class Solution:
    def calculate(self, s: str) -> int:
        
        s.replace(" ", "")
        toks = []
        token=""
        
        for c in s :
            if  c in ["+", "-", "*", "/",] :
                toks.append(token)
                token=""
                toks.append(str(c))
            else :
                token += str(c)
        
        toks.append(token)
        
        #1
        stk = []
        go = False
        prev = ""
        for op in toks :
            if go :
                go = False
                op2 = int(op)
                op1 = stk[-1]
                del stk[-1]
                if prev == "*" :
                    stk.append(op1*op2)
                else :
                    stk.append(op1//op2)
                continue

            if op in ["*", "/"] :
                go = True
                prev = op
                continue

            if op in ["+", "-"] :
                stk.append(op)
                continue

            stk.append(int(op))

        #2
        ret = 0
        prevPlus = True
        for op in stk :
            if op == "+" :
                prevPlus = True
                continue
            if op == "-" :
                prevPlus = False
                continue
                
            if prevPlus :
                ret += op
            else :
                ret -= op

        return ret

File: Google/059_BasicCalc2.cpp
// Calc2.cpp : https://leetcode.com/problems/basic-calculator-ii/

#include "stdafx.h"
#include <MyHeaders.h>

class Solution {
public:
	int calculate(string s)
	{
		stack<long> stk;
		set<char> signs = {'*', '/', '+', '-'};

		long lastnum = 0;
		char lastSign = '+';

		for (size_t i = 0; i < s.size(); i++)
		{
			char c = s[i];
			
			if ('0' <= c && c <= '9')
				lastnum = lastnum * 10 + c - '0';

			if (signs.find(c) == signs.end() && i < s.size() - 1)
				continue;

			if (lastSign == '+')
				stk.push(lastnum);
			else if (lastSign == '-')
				stk.push(-lastnum);
			else if (lastSign == '*')
			{
				lastnum = stk.top()*lastnum;
				stk.pop();
				stk.push(lastnum);
			}
			else if (lastSign == '/')
			{
				lastnum = stk.top()/lastnum;
				stk.pop();
				stk.push(lastnum);
			}

			lastnum = 0;
			lastSign = c;
		}

		int ret = 0;
		while (!stk.empty())
		{
			ret += stk.top();
			stk.pop();
		}

		return ret;
	}
};

int main()
{
	Solution sl;

	auto x = sl.calculate("3+2*2");
	PRN(x);
	
    return 0;
}

